### PR TITLE
Demo: Passing experiment data into Layout

### LINKED ISSUE #72889: Sidebar nudge upsell: Prepare the domain page to work with the experiment
[Based on our POC](https://github.com/Automattic/wp-calypso/pull/72831), we expect:

- [ ] We're not handling the page header here
- [ ] We're not handling the domain filter changes here
- [ ] Prepare the domain page to change its behavior if the user participates in the experiment
- [ ] Make sure we're implementing a good strategy to avoid changing the URL
- [ ] It should look like the design: pebzTe-Fx-p2#comment-1131
- [ ] Make sure the changes are connected to the experiment and only show when the user is in the `treatment` group. [Example of code here](https://github.com/Automattic/wp-calypso/pull/72831/commits/8148a79a089580437ddc43cbaa27d6ad03f381e0).

### Visual reference

Image

task-related: https://github.com/Automattic/wp-calypso/issues/72516

### PR DESCRIPTION
**Pull request by candy02058912**
#### Proposed Changes

Still works as the original PR https://github.com/Automattic/wp-calypso/pull/72932
Please check out the demo and testing instructions there.

* Instead of using `sessionStorage`, this PR loads the experiment data from an upper level and passes into <Layout />.
* Fixes: the error mentioned [here](https://github.com/Automattic/wp-calypso/pull/72932#issuecomment-1414511295)

#### Testing Instructions

* Please check out the demo and testing instructions in https://github.com/Automattic/wp-calypso/pull/72932

#### Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

Related to #

### CODE REVIEW

File: client/my-sites/domains/domain-search/index.jsx [hunk 1]
```diff
@@ -101,8 +103,18 @@ class DomainSearch extends Component {
...
+		const experimentAssignment = await loadExperimentAssignment(
+			'calypso_sidebar_upsell_dedicated_upgrade_flow_202301'
+		);
+		this.setState( {
+			isDomainSidebarExperimentUser: experimentAssignment?.variationName === 'treatment',
+		} );
+	};
+
+	async componentDidMount() {
+		await this.getExperimentData();
```
Comment (candy02058912) [inline client/my-sites/domains/domain-search/index.jsx:116]: IMO: the isDomainSidebarExperimentUser from sessionStorage is better here 🥲

File: client/my-sites/domains/index.js [hunk 2]
```diff
@@ -205,7 +204,6 @@ export default function () {
 	page(
 		'/domains/add/:domain',
 		siteSelection,
-		setExperimentVariation,
```
Comment (candy02058912) [inline client/my-sites/domains/index.js:208]: This was causing the SSR error

File: client/my-sites/controller.js [hunk 3]
```diff
@@ -428,16 +428,16 @@ export function noSite( context, next ) {
 	return next();
 }
 
-export async function setExperimentVariation( context, next ) {
-	const experimentAssignment = await loadExperimentAssignment(
-		'calypso_sidebar_upsell_dedicated_upgrade_flow_202301'
-	);
+// export async function setExperimentVariation( context, next ) {
```
Comment (tyxla) [inline client/my-sites/controller.js:431]: I'd suggest not committing commented code like this - we can always use code that's historically available by looking at older versions of the code and PRs.

File: client/controller/index.web.js [hunk 4]
```diff
@@ -39,9 +40,16 @@ export const ProviderWrappedLayout = ( {
...
+	const [ , experimentAssignment ] = useExperiment(
+		'calypso_sidebar_upsell_dedicated_upgrade_flow_202301'
+	);
 
 	const layout = userLoggedIn ? (
-		<Layout primary={ primary } secondary={ secondary } />
+		<Layout
+			primary={ primary }
+			secondary={ secondary }
+			experimentAssignment={ experimentAssignment }
```
Comment (tyxla) [inline client/controller/index.web.js:51]: Is `experimentAssignment` really necessary to provide this so early? Can we inject it where it's actually needed?

**Review by tyxla — COMMENTED** [review]
I realize this is merging into another feature branch, so I'm happy to skip a thorough review, but I'd strongly recommend that the commented code does not land in production unless we have a really good reason for that.

**Comment by paulopmt1** [comment]
We'll use [the previous PR](https://github.com/Automattic/wp-calypso/pull/72932) with suggestions from this one as our final version. Thanks for your help @candy02058912!

File: client/controller/index.web.js (hunk 4 above)
Comment (paulopmt1) [reply client/controller/index.web.js:51]: Good point @tyxla, [we fixed this behavior here](https://github.com/Automattic/wp-calypso/pull/72932/commits/8467502c2827e3d4a308f65763af209fe915a728)

File: client/my-sites/controller.js (hunk 3 above)
Comment (candy02058912) [reply client/my-sites/controller.js:431]: Thanks for the reminder, I agree with you, I was rushing myself to put something out in code to share my thoughts 😅

**Pull request closed without merging**